### ISSUE TITLE
[proposal] DataTable - have a way to disable the inkwell effect

### ISSUE THREAD

**Issue by nukotsuka** [flutter/flutter#70755]
## Problem

When I build `DataTable` with `DataRow.onSelectChanged == null` and `DataCell.onTap == null`, rendered cells in the table have `TableRowInkWell`.

## Causes

In `_buildDataCell`, if `onTap == null` and `onSelectChanged != null`, then `TableRowInkWell` is set to `label`.

https://github.com/flutter/flutter/blob/c1042314a94e693b4ea92404dab84852e9428442/packages/flutter/lib/src/material/data_table.dart#L837-L849

But this method is always called so that `onSelectChanged` is not null.

https://github.com/flutter/flutter/blob/c1042314a94e693b4ea92404dab84852e9428442/packages/flutter/lib/src/material/data_table.dart#L997

I think this part should look like this:

```dart
onSelectChanged: row.onSelectChanged != null ? () =>  row.onSelectChanged!(!row.selected) : null,
```

And I also found similar parts:
https://github.com/flutter/flutter/blob/c1042314a94e693b4ea92404dab84852e9428442/packages/flutter/lib/src/material/data_table.dart#L937
https://github.com/flutter/flutter/blob/c1042314a94e693b4ea92404dab84852e9428442/packages/flutter/lib/src/material/data_table.dart#L724-L730

## Flutter Doctor
<details>
  <summary>Log</summary>

```
[✓] Flutter (Channel beta, 1.23.0-18.1.pre, on Mac OS X 10.15.7 19H2 x86_64, locale ja-JP)
    • Flutter version 1.23.0-18.1.pre at /Users/taka/Flutter/flutter
    • Framework revision 198df796aa (5 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/taka/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/taka/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 41.1.3
    • Dart plugin version 192.7402

[✓] Connected device (3 available)
    • nukotsuka 11 (mobile) • 00008030-001108E20AD0802E • ios            • iOS 14.0
    • Web Server (web)      • web-server                • web-javascript • Flutter Tools
    • Chrome (web)          • chrome                    • web-javascript • Google Chrome 86.0.4240.198

• No issues found!
```

</details>

**Comment by pedromassangocode** [comment]
Hi @nukotsuka 
Can you elaborate what you are trying to achieve?
Are you proposing to have a way to disable the inkwell effect?

**Comment by nukotsuka** [comment]
@pedromassangocode 
I'm trying to achieve `DataTable` with no inkwell effect in web.
So I want to disable the effect both on hover and clicked so that users don't misunderstand that cells are clickable.
As you say, I'm proposing to have a way to disable the inkwell effect.

**Comment by laureljanej** [comment]
@pedromassangocode This issue also has an effect on users of screen readers. Similarly to the visual inkwell effect occurring when there is no action associated with a row, Android TalkBack announces "double tap to activate" which is misleading and can cause confusion.  

The fix that @nukotsuka identified should also fix the screen reader issue but I wanted to put the accessibility issue on your radar as well. Thank you!

**Comment by digit81** [comment]
As workaround  for lack of nonactive rows I wrapped whole `DataTable` with `IgnorePointer`.
It will also turn off sorting, but in my case I don't need header at all.

Unfortunately it also lacks hover effect on rows - which I would like to have.

```dart
IgnorePointer(
  child: DataTable(
    headingRowHeight: 0,
    showCheckboxColumn: false,
    columns: [DataColumn(label: Container()), DataColumn(label: Container())],
    rows: [...]
  ),
);
```

**Comment by buimanhhungit** [comment]
![image](https://user-images.githubusercontent.com/51559988/126854357-89601f22-ec0f-4f2c-835c-9259fdfbf949.png)

**Comment by rshrc** [comment]
> ```
>  DataRow(
>                         color: MaterialStateProperty.resolveWith<Color>((Set<MaterialState> states) {
>                           if (states.contains(MaterialState.selected))
>                             return Theme.of(context).colorScheme.primary.withOpacity(0.08);
>                           return Colors.transparent;  
>                         }),
> ```

This fix worked just as expected!! You are the messiah!! ;-;

**Comment by rignaneseleo** [comment]
I think my case is related to this: I am working with Flutter web and the mouse cursor turns into a hand even though both `onLongPress` and `onSelectChanged` are explicitly null.

**Comment by TahaTesser** [comment]
> I think my case is related to this: I am working with Flutter web and the mouse cursor turns into a hand even though both onLongPress and onSelectChanged in DataRow are explicitly null.


I don't see the InkWell effect when both `onLongPress` and `onSelectChanged` are null. 


> So I want to disable the effect both on hover and clicked so that users don't misunderstand that cells are clickable.
> As you say, I'm proposing to have a way to disable the inkwell effect.

> I can't use IgnorePointer on the DataTable because I have a button on each row that should be clickable and show the hand cursor.

Added ability to only select row with the checkbox with no ink well and `IgnorePointer` won't be needed any buttons will be clickable when they're part of the row.

**Comment by rignaneseleo** [comment]
I'm not talking about the inkwell but about mouse cursor turning into a hand even though the events are set null.

**Comment by TahaTesser** [comment]
> I'm not talking about the inkwell but about mouse cursor turning into a hand even though the events are set null.T

Thanks, since there is no inkwell without events, I don't see the mouse cursor updating on the web.
Can you please provide your `flutter doctor -v` output and a minimal code sample?

**Comment by TahaTesser** [comment]
> In _buildDataCell, if onTap == null and onSelectChanged != null, then TableRowInkWell is set to label.

> @pedromassangocode This issue also has an effect on users of screen readers. Similarly to the visual inkwell effect occurring when there is no action associated with a row, Android TalkBack announces "double tap to activate" which is misleading and can cause confusion.

https://github.com/flutter/flutter/pull/105123 adds ability to avoid inkwell when `onSelectChanged != null,`  and this issue is closed.


> I'm not talking about the inkwell but about mouse cursor turning into a hand even though the events are set null.


Please file a separate issue if this issue persists.
